schemas/schema-core/src/main/java/org/fao/geonet/kernel/schema/subtemplate/SubtemplatesByLocalXLinksReplacer.java

```diff
@@ -0,0 +1,111 @@
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ *
+ * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
+ * Rome - Italy. email: geonetwork@osgeo.org
+ */
+
+package org.fao.geonet.kernel.schema.subtemplate;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.fao.geonet.index.es.EsRestClient;
+import org.fao.geonet.utils.Xml;
+import org.jdom.Element;
+import org.jdom.Namespace;
+
+public abstract class SubtemplatesByLocalXLinksReplacer  {
+    private Map<String, Replacer> replacersDict = new HashMap<>();
+    private ManagersProxy managersProxy;
+    private List<Namespace> namespaces;
+    private EsRestClient esRestClient;
+
+    public SubtemplatesByLocalXLinksReplacer(List<Namespace> namespaces, ManagersProxy managersProxy, EsRestClient esRestClient) {
+        this.managersProxy = managersProxy;
+        this.namespaces = namespaces;
+        this.esRestClient = esRestClient;
+    }
+
+    public static String FORMAT = "format";
+    public static String CONTACT = "contact";
+    public static String EXTENT = "extent";
+    public static String KEYWORD = "keyword";
+
+    public Element replaceSubtemplatesByLocalXLinks(Element dataXml,
+                                                    String templatesToOperateOn) {
+
+        String localXlinkUrlPrefix = "local://srv/api/registries/entries/";
+        final String lang = computeMetadataLang(dataXml);
+        final List<String> localesAsHrefParam = getLocalesAsHrefParam(dataXml);
+        final String localisedCharacterStringLanguageCode =
+                String.format("#%S", managersProxy.getIso1LangCode(lang));
+
+        Status status = Arrays.stream(templatesToOperateOn.split(";"))
+                .filter(replacersDict::containsKey)
+                .map(replacersDict::get)
+                .map(replacer -> replacer.replaceAll(dataXml,
+                        localXlinkUrlPrefix,
+                        esRestClient,
+                        localisedCharacterStringLanguageCode,
+                        lang,
+                        localesAsHrefParam))
+                .collect(Status.STATUS_COLLECTOR);
+
+        if (status.isError()) {
+            throw new RuntimeException(status.msg);
+        } else {
+            return dataXml;
+        }
+    }
+
+    public void addReplacer(Replacer replacer) {
+        replacersDict.put(replacer.getAlias(), replacer);
+    }
+
+    private String computeMetadataLang(Element dataXml) {
+        String lang = "eng";
+        try {
+            lang = Xml.selectElement(dataXml, ".//gmd:language/gmd:LanguageCode", namespaces).getAttributeValue("codeListValue");
```

Comment:
This is a bit to ISO19139 focus, we should consider using extract-default-language.xsl. So we only search subtemplate in the main metadata language right ?